nullaway/src/main/java/com/uber/nullaway/NullAway.java

```diff
@@ -437,6 +456,232 @@ public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState
+    }
+    if (symbolHasSuppressInitalizationWarningsAnnotation(symbol)) {
+      // also suppress checking read before init, as we may not find explicit initialization
+      return Description.NO_MATCH;
+    }
+    return checkPossibleUninitFieldRead(
+        tree, state, symbol, path, enclosingBlockPath, enclosingClassPath);
+  }
+
+  private boolean relevantInitializerMethodOrBlock(Tree methodLambdaOrBlock, VisitorState state) {
+    if (methodLambdaOrBlock instanceof LambdaExpressionTree) {
+      return false;
+    } else if (methodLambdaOrBlock instanceof MethodTree) {
+      MethodTree methodTree = (MethodTree) methodLambdaOrBlock;
+      return isConstructor(methodTree) && !constructorInvokesAnother(methodTree, state);
+    } else {
+      // initializer or field declaration
+      return true;
+    }
+  }
+
+  private Description checkPossibleUninitFieldRead(
+      IdentifierTree tree,
+      VisitorState state,
+      Symbol symbol,
+      TreePath path,
+      TreePath enclosingBlockPath,
+      TreePath enclosingClassPath) {
+    if (!fieldInitializedByPreviousInitializer(
+            symbol, enclosingBlockPath, enclosingClassPath, state)
+        && !fieldAlwaysInitializedBeforeRead(
+            symbol, path, state, enclosingBlockPath, enclosingClassPath)) {
+      return createErrorDescription(
+          MessageTypes.NONNULL_FIELD_READ_BEFORE_INIT,
+          tree,
+          "read of @NonNull field " + symbol + " before initialization",
+          path);
+    } else {
+      return Description.NO_MATCH;
+    }
+  }
+
+  /**
+   * @param symbol the field being read
+   * @param pathToRead TreePath to the read operation
+   * @param state visitor state
+   * @param enclosingBlockPath TreePath to enclosing initializer block
+   * @param enclosingClassPath TreePath to enclosing class
+   * @return true if within the initializer, the field is always initialized before the read
+   *     operation, false otherwise
+   */
+  private boolean fieldAlwaysInitializedBeforeRead(
+      Symbol symbol,
+      TreePath pathToRead,
+      VisitorState state,
+      TreePath enclosingBlockPath,
+      TreePath enclosingClassPath) {
+    AccessPathNullnessAnalysis nullnessAnalysis = getNullnessAnalysis(state);
+    Set<Element> nonnullFields;
+    if (symbol.isStatic()) {
+      nonnullFields = nullnessAnalysis.getNonnullStaticFieldsBefore(pathToRead, state.context);
+    } else {
+      nonnullFields = new LinkedHashSet<>();
+      nonnullFields.addAll(
+          nullnessAnalysis.getNonnullFieldsOfReceiverBefore(pathToRead, state.context));
+      nonnullFields.addAll(
+          safeInitByCalleeBefore(pathToRead, state, enclosingBlockPath, enclosingClassPath));
+    }
+    return nonnullFields.contains(symbol);
+  }
+
+  /**
+   * computes those fields always initialized by callee safe init methods before a read operation
+   * (pathToRead) is invoked
+   */
+  private Set<Element> safeInitByCalleeBefore(
```

Comment:
I am not 100% sure what this method does or why it makes sense only for instance initializers and not for static initializers. Could we expand on the documentation here a little bit?

Is this just checking that if I do `m1(); c = this.f;` and `m1()` writes to `this.f` that counts for initialization? Or is it doing something more?

Also, more generally: We are assuming that consumers will call some `@Initializer` method before calling any other public methods for the class/instance, and just checking here that intializer blocks, constructors and `@Initializer` methods, and the private methods they call, don't use fields before initialization. Correct? We are not trying to detect if consumers are failing to call the `@Initializer` methods. Right?

Reply:
So for instance fields, we are trying to match the checking documented [here](https://github.com/uber/NullAway/wiki/Error-Messages#initializer-method-does-not-guarantee-nonnull-field-is-initialized--nonnull-field--not-initialized).  We allow instance fields to be initialized in callees of constructors / initializer methods, provided certain conditions are met (see the docs).  Since we allow this type of initialization, we definitely don't want to incorrectly warn about reading before init in such cases.

As to why we don't allow similar initialization patterns for static fields, it's just that we never implemented that, as we never saw a need.  We can do that in the future, but I think that should be separate from this change.

Right now, we are doing *no* checking of `@Initializer` methods.  As a follow up, I think we can check that if there is exactly one `@Initializer` method, it does not read any field before writing it if that field may not have been initialized in the constructor(s).  That's probably a good thing to implement, actually.   I can do that in a follow-up PR.

Also, you're correct that we are not doing any checking that `@Initializer` methods are invoked before other methods in the class.

Reply:
Sounds good to me, maybe linking the documentation in a comment is a good idea here?